「lsコマンドを作る2」の課題である-aオプションが使えるlsコマンドを実装

cellotak opened the pull request:
[lsコマンドを作る2 \| FBC](https://bootcamp.fjord.jp/practices/222)の課題として、lsコマンドに-aオプションを実装しました。

コードの中での補足や気になる点はコメントで記載しています。

レビューをお願い致します。


## 実行結果
※提出フォームでも記載していますが、PRだけ見られる方もいるので、実行結果をdescriptionにも記載します。

### 作成したlsコマンドの実行結果


![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOHdqQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--db56e43d7b925743816c44a22a78500f364b3b67/image.png)



上から順に
1. オプションなし、ディレクトリの指定なし
2.  `-a`オプション有、ディレクトリ指定なし
3.  オプションなし、ディレクトリ指定有
4. `-a`オプション有、ディレクトリ指定有

という条件で実行しています。

### OS標準のlsコマンドの実行結果

![image.png](https://bootcamp.fjord.jp/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBODBqQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--973c777509d04d44f3b062d3ad2db54f13ddd942/image.png)




同じく
1. オプションなし、ディレクトリの指定なし
2.  `-a`オプション有、ディレクトリ指定なし
3.  オプションなし、ディレクトリ指定有
4. `-a`オプション有、ディレクトリ指定有

という条件で実行しています。

### その他補足
- 平仮名や漢字にすると、左揃えでなくなる問題にはまだ手をつけていません
- ディレクトリ名は指定できますが、ファイル名を指定することはまだできません。(今後どこかのタイミングで対応したいとは思っています。)
- lsコマンドにaliasを設定しており、`ls`と打つとディレクトリ名は青色になるなど、標準とは異なる表示がされてしまうので、`\ls`という形で実行しています。
- OS標準のlsコマンドとはソートのルールが違うようで、出力順は異なっています

cellotak committed:
ls.rbに-aオプションを実装

ただし、-aオプションの後にディレクトリ指定すると動作しなくなるため、修正が必要

cellotak committed:
オプションの位置とディレクトリ名指定の位置を自由にできるように変更

cellotak committed:
ls.rbのrubocopでの指摘点を修正

cellotak commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,45 @@
 end
 
+option = search_option
 target_path = ARGV[0] || './'
```
現在コマンドライン引数でディレクトリ指定した際に、`ARGV[0]`をそのディレクトリ名として使用しているため、オプション自体に引数がある場合や、そのほかに引数があった場合に対応できません。
しかしまだオプションを増やしていく段階ですので、ここはこのままにして最終的に対応できるように変更したいと考えています。

kazzbps replied:
今回 optparse を採用したということなとので、 optparse の基本的な使い方をもう少し把握されるとよいと思います

https://docs.ruby-lang.org/ja/latest/library/optparse.html
の概要

![image](https://github.com/cellotak/ruby-practices/assets/24219290/b5eac2cc-98e9-45d9-9deb-59c91edb08bf)


`OptionParser#parse!` の戻り値は、オプションを除いた要素になり、引数に指定したファイル名(ディレクトリ名)を取得できることになります

optparse は、引数のオプションのみを得るのではなく、引数の非オプションと指定されたオプションの両方を取得できるライブラリになるかと思います

ということで、optparse を使うと `ARGV[0]` ではなく、
parse!(ARGV) でオプションを含まない指定されたファイル(ディレクトリ)名が得られるようになると思いますので
オプションの追加より、複数ファイル指定ができない等の問題を解決するのがよいかなと思います

cellotak replied:
@kazzbps 
もしかすると少し行き違いになっているかもしれませんが、
既に13行目で`opt.parse!(ARGV)`を使ってARGVからオプションを取り除けています。
ですのでオプションおよび1つ以下のディレクトリ指定があった場合は順不同で動作します。

ただし、オプションに紐づいた引数 (例えば`-T`オプションはタブ幅を8の代わりに引数の数字にする)　があった場合や、オプション以外の引数が2つ以上あった場合には現在対応していない、という意味で、この説明をしていました。

もし的はずれなコメントでしたらすみません。　

kazzbps replied:
コメントありがとうございます。

43行目の `ARGV[0]` は search_options の副作用でオプション引数が消えているので、引数に最初に指定されたディレクトリ名となり、ディレクトリ名の複数指定は今はやらないと理解しました

`target_path` という変数名ですが、`current_path` の方がニュアンスが近いのかもと思いました
※ ls は ファイル名も引数にとれますが、ファイル名の引数指定は未対応という理解です

cellotak replied:
> target_path という変数名ですが、current_path の方がニュアンスが近いのかもと思いました

ディレクトリ名を指定した際はcurrentディレクトリではないpath名が入るのですが、その点踏まえても`current_path`の方が合っていますでしょうか。


> ※ ls は ファイル名も引数にとれますが、ファイル名の引数指定は未対応という理解です

その通りです。

cellotak commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,45 @@
   filenames_table = filenames.each_slice(number_of_row).to_a
 
-  widths = filenames_table.map { |filenames| filenames.map(&:size).max + SPACE_WIDTH }
+  widths = filenames_table.map { |col| col.map(&:size).max + SPACE_WIDTH }
```
引数で引いてきた`filenames`と
`filenames_table`の中身としての`filenames`
が名前被りしていたので後者を`col` と変更しました。
実態は列の要素そのものなので`col`で問題ないと考えました。
前回は 35行目の処理の部分で
```ruby
filenames_table[col][row]
```
としていて、列そのものの配列としての`col`と、列番号を表す数値としてのcolが同じ名前でわかりづらいという指摘を頂いたので

```ruby
filenames_table[col_index][row_index]
```
として、列番号としての`col`は`col_index`とすることで混乱を避けるようにしてあります。

cellotak committed:
ls.rbの変数名を修正

kazzbps commented:
PRのdescriptionには、今回実装した機能を実行するとどうなるかを書くとよいと思います

* `./ls -a`

```
.   .gitignore  ls.rb  
..  .gitkeep    
```

* `./ls -a ..`

```
.             01.fizzbuzz  06.bowling_object  
..            02.calendar  07.ls_object       
.git          03.bowling   98.rake            
.gitignore    04.ls        99.wc_object       
.rubocop.yml  05.wc        README.md        
```

cellotak commented:
> PRのdescriptionには、今回実装した機能を実行するとどうなるかを書くとよいと思います
> 
> * `./ls -a`
> 
> ```
> .   .gitignore  ls.rb  
> ..  .gitkeep    
> ```
> 
> * `./ls -a ..`
> 
> ```
> .             01.fizzbuzz  06.bowling_object  
> ..            02.calendar  07.ls_object       
> .git          03.bowling   98.rake            
> .gitignore    04.ls        99.wc_object       
> .rubocop.yml  05.wc        README.md        
> ```

cellotak commented:
@kazzbps 

descriptionについて承知しました。
一応FBCの提出フォームの方で、スクリーンショットともに機能説明をしていますが、PRだけ見る方も想定して書こうと思います。

kazzbps commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,45 @@
     NUMBER_OF_COL.times do |col_index|
       # filenames_tableの行と列が逆で保存されているので、col_indexとrow_indexを入れ替えて出力させている
-      target_filename = filenames_table[col_index][row_index] 
+      target_filename = filenames_table[col_index][row_index]
```
filenames が 3(=NUMBER_OF_COL) を下回る場合にエラーが出ると思われます

具体的には以下のようになります

```
> ls one
1.txt

> ./ls.rb one
1.txt  ./ls.rb:35:in `block (2 levels) in output': undefined method `[]' for nil:NilClass (NoMethodError)

      target_filename = filenames_table[col_index][row_index]
                                                  ^^^^^^^^^^^
```

cellotak replied:
バグを見つけていただきありがとうございます。
filenames が 3(=NUMBER_OF_COL) を下回る場合でも表示できるように修正しました。

kazzbps replied:
修正確認しました 👍

kazzbps commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,45 @@
 def output(filenames)
   number_of_row = ((filenames.size - 1) / NUMBER_OF_COL) + 1
   # NOTE: filenames_tableの要素は行と列が出力したい形とは逆で保存されている
   filenames_table = filenames.each_slice(number_of_row).to_a
```
filenames が 空配列の場合エラーが発生すると思います

具体的には空ディレクトリを指摘すると以下のようになります

```
> ls empty

> ./ls.rb empty
./ls.rb:28:in `each_slice': invalid slice size (ArgumentError)
```

cellotak replied:
こちらもご指摘ありがとうございます。
filenamesが空の時はoutputの処理を行わないという方法でエラーを回避するよう修正しました。

kazzbps commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,45 @@
 def output(filenames)
   number_of_row = ((filenames.size - 1) / NUMBER_OF_COL) + 1
   # NOTE: filenames_tableの要素は行と列が出力したい形とは逆で保存されている
   filenames_table = filenames.each_slice(number_of_row).to_a
```
このコメントの意図がつかみかねていましたが、 lsの折り返しは転置されるのですね
(恥ずかしながら知らなかった...

```
> ls eight
1.txt    4444.txt    7777777.txt
22.txt   55555.txt   88888888.txt
333.txt  666666.txt
```

上記を each_slice(3) すると

```ruby
[
  [      "1.txt",       "22.txt",    "333.txt"],
  [   "4444.txt",    "55555.txt", "666666.txt"], 
  ["7777777.txt", "88888888.txt"]
]
```

となるので、これを行ごと(上から)ではなく列ごと(左から)に読む戦略と理解しました

cellotak replied:
いきなり見ると、lsコマンドが転置して表示する仕様になっていることに気づきづらいですね。
その背景がわかるように一応コメント追記しておきました。

kazzbps reviewed (changes requested):
ファイル数の少ないディレクトリを指定するとエラーが出るようなのでご確認お願いします

cellotak committed:
ls.rbでファイル名が2つ以下の場合にエラーとなるバグを修正

ファイル名が1or2の時、col_indexが2や3になっていて、filenames_tableが参照できていなかったため修正
ファイル名が0のとき、number_of_rowが0となりeach_sliceができていなかったため修正

cellotak committed:
ls.rbのコメントと変数名を修正

cellotak committed:
ls.rbの変数名を修正

cellotak committed:
コメントを追加

kazzbps commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,51 @@
-filenames = get_filenames(target_path)
-output(filenames)
+filenames = get_filenames(target_path, option_params)
+output(filenames) unless filenames.empty?
```
空フォルダ指定時のエラー修正確認しました。
このロジックは output メソッドの中でやるとよいと思います

outputメソッドの仕様は
```
outputメソッドは引数に配列を取る。
ただし1件以上要素を必須とし、空配列を渡さないこと(呼び出し時にチェックすること)
```

というなりますが、

```
outputメソッドは引数に配列を取る。
要素数が0件の場合は出力を行わない(呼び出し時に要素数チェックは不要)
```

となる方がoutputメソッドを再利用しやすくなります

具体的にはガード構文が使えると思います
https://qiita.com/kouyan/items/7b8b456b626447a1e24e

cellotak replied:
空フォルダ指定時の処理をoutputメソッド内に変更しました。

kazzbps reviewed (approved):
バグ修正確認しました

空配列のガードに関して見直してみましょう

morimorihoge commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,51 @@
 
-def get_filenames(target_path)
-  Dir.glob('*', base: target_path)
+def search_option
```
COMMENT:

こちらのメソッドですが、引数に `ARGV` を撮る用に作っておくと、このメソッド自体が独立した実装になりテストしやすくなるかなと思います。

このあたりは後のリファクタリングでやればいい話でもありますが、testableなメソッドの意識をできると、機能分割の目安になるかなあと思いました。

cellotak replied:
引数にARGVをとれるようにしました。

JunichiIto replied:
これ、細かいことを言うと、`parse!`で引数の内容を破壊的に変えちゃうんですよね。
OptionParserやARGVを使うなら特殊なケースなので許容範囲か、という気もするのですが、できれば非破壊的な`parse`を使って引数を変更しない方が一般的なメソッドの設計としてお行儀がいいような気がします。

[\!が付く破壊的メソッドの使用は原則として避ける \| FBC](https://bootcamp.fjord.jp/pages/do-not-use-destructive-method)

cellotak replied:
下記のようにopt.parseの戻り値をローカル変数に入れて、2つまとめて戻り値にするように変更しました。受け取り先では多重代入になってます。

```ruby
  arguments_without_options = opt.parse(argv)
  return option_params, arguments_without_options
```

しかし、rubocopには一つの配列に入れるように指摘されてます。
個人的には戻り値を2つにして、受取先で多重代入する書き方の方がしっくり来たのですが、どうすべきか悩んでます。
![image](https://github.com/cellotak/ruby-practices/assets/65953037/85f0b30d-b4ee-4367-afc8-a9b5d565df08)


それとも、そもそも`search_option`メソッドの役割が2つになっているので、それを分けるべきでしょうか。
その場合、ARGVからoptionを取り出すメソッド、ARGVからoption除いてディレクトリパスを取り出すメソッド、の2つにわける方法として
1. opt.parseを2回繰り返すことを許容して、メソッドを分ける
1.  クラスを作ってインスタンス変数
といったものが考えられますが、どちらが良いと思われるでしょうか。あるいは他の案などありますでしょうか。

長くなってしまいましたがご確認お願い致します。

cellotak committed:
ls.rbのserch_optionメソッドが引数を受け取れるようにしました

cellotak committed:
ls.rbの空フォルダ指定時の処理をoutputメソッドに移動させました

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,53 @@
+  if params[:a]
+    Dir.entries(target_path).sort
+  else
+    Dir.glob('*', base: target_path)
```
片方はentries + sort、もう片方はglobのみ、で対称性がなくてちょっと気持ち悪いですね。
メソッドが違うので異なるデータが返ってきそうな印象を受けます。
なるべくどちらも同じメソッドを使うように書き替えることはできますか？

cellotak replied:
もう一度公式リファレンスを見直したら、globでも隠しファイルをマッチさせる方法をみつけたので、globに統一しました。
しかし、[リビジョン 5917f49a \- Fix documentation for Dir\.glob Dir\.glob\("\*", File::FNM\_DOTMATCH\) no longer i\.\.\. \- Ruby master \- Ruby Issue Tracking System](https://bugs.ruby-lang.org/projects/ruby-master/repository/git/revisions/5917f49ad03cd3b2c13d3eea85c969a113b83bf8)
にもある通り、Ruby3.1 以降では`File::FNM\_DOTMATCH`というフラグを設定しても
親ディレクトリを表す`..`にマッチしないため、OS標準のlsコマンドとは出力が異なっています。

ただ今回は[lsコマンドを作る2 ](https://bootcamp.fjord.jp/practices/222#-4)の注意事項に
> .. については出力されていなくても問題ありません。

とあったため、この状態で提出させていただきます。

なお、sortについてはglobのパラメータとして設定できることに気が付いたため、その方法で統一しています。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,53 @@
       # filenames_tableの行と列が逆で保存されているので、col_indexとrow_indexを入れ替えて出力させている
-      target_filename = filenames_table[col_index][row_index] 
+      target_filename = filenames_table[col_index][row_index]
       print target_filename.ljust(widths[col_index]) unless target_filename.nil?
```
```suggestion
      print target_filename&.ljust(widths[col_index])
```

でもいいかも。

cellotak replied:
チェリー本で出てきたぼっち演算子、ここで使えたんですね！
その部分、もやもやするなと思っていましたが非常にスッキリしました。
ありがとうございます！

cellotak committed:
ls.rbのファイル名取得の記述をリファクタリングしました

cellotak committed:
ls.rbのnilの判定を&.演算子に変更

cellotak committed:
ls.rbでopt.parseの破壊的メソッドの仕様を避けるようにリファクタリングしました

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,53 @@
+  option_params = {}
+  opt.on('-a') { |v| option_params[:a] = v }
+  arguments_without_options = opt.parse(argv)
+  return option_params, arguments_without_options
```
```suggestion
  [option_params, arguments_without_options]
```

rubocopの指摘はこれで直るはずですー。

cellotak replied:
ありがとうございます。とりあえず配列で値を返すように変更しました。
https://github.com/cellotak/ruby-practices/pull/7#discussion_r1286663733
の後半で相談させて頂いた内容についてはいかがでしょうか。(メソッドを分けるべきかどうか)

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,53 @@
+  opt = OptionParser.new
+  option_params = {}
+  opt.on('-a') { |v| option_params[:a] = v }
+  arguments_without_options = opt.parse(argv)
```
この変数名、長い上にあまり抽象化できてないですね。。
オプションを取り除いたあとに残る情報はファイルパスになるはずなので、

```suggestion
  paths = opt.parse(argv)
```

とか

```suggestion
  file_paths = opt.parse(argv)
```

みたいな名前の方がいいと思います。

cellotak replied:
なるほど、確かに言われてみるとわかりづらい名前ですね。
変数名の付け方がまだまだだと感じます。
提案頂いた`file_paths`に変更させて頂きます。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,53 @@
+  if params[:a]
+    Dir.glob('*', File::FNM_DOTMATCH, base: target_path, sort: true)
+  else
+    Dir.glob('*', base: target_path, sort: true)
```
`File::FNM_DOTMATCH`以外は同じですね。
オプションなしはたしか整数の0だったと思うので、`params[:a]`の有無で第2引数を切り替えるようにして、それ以外はDRYにしたい（globを使うのは一回だけにしたい）です。

あと、sortオプションはデフォルトでtrueになってるはずなので省略してもいいかなと思いました。

cellotak replied:
> File::FNM_DOTMATCH以外は同じですね。

気付いてませんでした。globを使うのは1回だけになるよう`File::FNM_DOTMATCH`の部分を変数にしました。ここでも3項演算子が使えることに気が付いたのでスッキリさせることができました。

sortオプションについて、デフォルトの設定に気づいていませんでした。省略しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,53 @@
 
-  widths = filenames_table.map { |filenames| filenames.map(&:size).max + SPACE_WIDTH }
+  number_of_row = ((filenames.size - 1) / NUMBER_OF_COL_MAX) + 1
+  number_of_col =
+    if filenames.size < NUMBER_OF_COL_MAX
+      filenames.size
+    else
+      NUMBER_OF_COL_MAX
+    end
```
好みの問題かもしれませんが、三項演算子で1行にまとめたい。。

cellotak replied:
3項演算子に変更しました。
こういった場面で使えるのですね。

cellotak replied:
https://github.com/cellotak/ruby-practices/pull/7#discussion_r1288109178 のコメントを受け、二つのminを取る方法にさらに変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,53 @@
 
-target_path = ARGV[0] || './'
-filenames = get_filenames(target_path)
+option_params, arguments_without_options = search_option(ARGV)
```
option_paramsはoptionなのかparamsなのか、どっちかに寄せてしまっていいと思います。
`def get_filenames(target_path, params)` の引数名と合わせる、という方針がいいかもしれません。

参考 [ユーザーを見つけたら……Aさん「自転車に入れます」Bさん「人間に入れます」Cさん「ユーザーに入れます」 \| FBC](https://bootcamp.fjord.jp/pages/readable-variable-name)

cellotak replied:
なるほど、おっしゃるとおりその部分が統一できていませんでしたね。
`params`だとなんのパラメータかわかりづらい気がしたので、今回はコード全体を通して`options`に変更しました。

JunichiIto reviewed (commented):
コメントしましたー。

cellotak committed:
ls.rbのsearch_optionの変数名を変更

cellotak committed:
ls.rbのトップレベルの変数名を変更

cellotak committed:
ls.rbの記述ミスを修正

cellotak committed:
ls.rbの変数名変更

cellotak committed:
ls.rbのoutputメソッドのif文を3項演算子に変更

cellotak committed:
ls.rbの変数paramsをoptionsに変更

cellotak committed:
ls.rbのget_filenamesをリファクタリング

cellotak committed:
ls.rbのsearch_optionメソッドの戻り値を複数の変数から単一の配列に変更

ebihara-tomonari commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,46 @@
+  return if filenames.empty?
+
+  number_of_row = ((filenames.size - 1) / NUMBER_OF_COL_MAX) + 1
+  number_of_col = filenames.size < NUMBER_OF_COL_MAX ? filenames.size : NUMBER_OF_COL_MAX
```
ここ三項演算子で行数減らしちゃってますが、実は意味合い的に考えると微妙です
```ruby
number_of_col = [filenames.size, NUMBER_OF_COL_MAX].min
```
とするのが良さげです

`dotmatch_flag = options[:a] ? File::FNM_DOTMATCH : 0` この三項演算子の使い方なら良い 👌

cellotak replied:
ありがとうございます！それは目から鱗でした！

cellotak committed:
ls.rbのoutputメソッドのリファクタリング

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -47,7 +40,7 @@ def output(filenames)
-option_params, arguments_without_options = search_option(ARGV)
-target_path = arguments_without_options[0] || './'
-filenames = get_filenames(target_path, option_params)
+options, file_paths = search_option(ARGV)
```
searchよりもparseの方がメソッドの役割に即しているように思います。あとoptionかoptionsかも揃えた方がベターですね。

```suggestion
options, file_paths = parse_options(ARGV)
```

cellotak replied:
searchからparseに変更しました。

optionのその部分漏れてました！修正しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,46 @@
-NUMBER_OF_COL = 3
+require 'optparse'
+
+NUMBER_OF_COL_MAX = 3
```
プログラミングの変数名やメソッド名ってあまりofは使われない気がします。（ゼロではないけどあまり見ない）
なので、 b_of_a みたいな変数名は a_b みたいに書いた方が馴染みが良いです。
あと、maxを使うなら、a_max よりも max_a の方が自然かも。
あとnumberは「番号」のように解釈される可能性もありそう。ここはcountとかの方がいいかな。

・・・という見解を踏まえて新しい定数の名前を考えてみてほしいです！

cellotak replied:
命名について解説頂きありがとうございます。
「××の数」を表す変数を調べたところ、`××_count`とするパターンが多そうだったため、

`NUMBER_OF_COL_MAX` ⇒ `MAX_COL_COUNT` としてみました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -1,30 +1,46 @@
-  filenames_table = filenames.each_slice(number_of_row).to_a
+  return if filenames.empty?
+
+  number_of_row = ((filenames.size - 1) / NUMBER_OF_COL_MAX) + 1
```
ここも上の定数の指摘と同じです。

cellotak replied:
同様に、
 `number_of_row` ⇒ `row_count` 
 `number_of_col`  ⇒  `col_count`
 のように変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -47,7 +40,7 @@ def output(filenames)
-filenames = get_filenames(target_path, option_params)
+options, file_paths = search_option(ARGV)
+target_path = file_paths[0] || './'
+filenames = get_filenames(target_path, options)
```
Rubyって`get_`や`set_`っていうprefixはあまり使わないので（これもゼロではないです）、別の動詞を使った方がいいかも。
それこそ`search_`とか、`fetch_`とか。

あと、めっちゃ細かいんですが、 filename とto file_path でアンダーバーの使い方に一貫性がないのも気になります。。

cellotak replied:
`fetch`の方がイメージに近いと思ったので、`fetch`に変更しました。

アンダーバーの一貫性は気づいていませんでした！
`filename`や`filenames`はアンダーバーで分けることにしました。

JunichiIto reviewed (commented):
コメントしました〜。

> それとも、そもそもsearch_optionメソッドの役割が2つになっているので、それを分けるべきでしょうか。

これぐらいなら許容範囲かな〜と思います。
起動時引数を処理した結果、2つのデータを返す必要があった、という認識です。

こだわるなら、ハッシュや専用クラスを作ってそれを返す、とかですかね〜。

```ruby
def search_option(argv)
  opt = OptionParser.new
  options = {}
  opt.on('-a') { |v| options[:a] = v }
  file_paths = opt.parse(argv)
  { options: options, file_paths: file_paths }

  # もしくは
  WcArgs.new(options, file_paths)
end
```

cellotak committed:
各種命名を変更

cellotak commented:
> これぐらいなら許容範囲かな〜と思います。
> 起動時引数を処理した結果、2つのデータを返す必要があった、という認識です。
> こだわるなら、ハッシュや専用クラスを作ってそれを返す、とかですかね〜。

@JunichiIto 
分かりました！ここからまたロジックを書き換えるのは時間かかりすぎてしまいそうなので、今回は一旦このままで進めたいと思います。
(またlsコマンドの最後で挑戦させてもらうかもしれません。)

cellotak commented on 04.ls/ls.rb:
```diff
@@ -3,44 +3,45 @@
+options, directory_paths = parse_options(ARGV)
+target_directory_path = directory_paths[0] || './'
+file_names = fetch_file_names(target_directory_path, options)
+output(file_names)
```
レビューして頂いた点とは別になりますが、
元々`file_paths`としていた変数はファイルのパスではなく、「名前を表示させたいファイルがあるディレクトリのパス」が入っている状態で不自然に感じたので、`directory_paths`という風に変更しました。

また`target_path`もあくまでもディレクトリのパスなので同様に`target_directory_path`に変更しました。

JunichiIto commented on 04.ls/ls.rb:
```diff
@@ -3,44 +3,45 @@
-output(filenames)
+# directory_pathsには複数のpathを指定することは許容しているが、現時点でファイル名を表示するのは1番目に指定したディレクトリのみにしている。
+options, directory_paths = parse_options(ARGV)
+target_directory_path = directory_paths[0] || './'
```
targetじゃないdirectory_pathがあるのか？と言ったらないと思うので、

```suggestion
directory_path = directory_paths[0] || './'
```

でもいいんじゃないかなーと思います。

cellotak replied:
たしかにtargetじゃないものはないですね
修正しておきます！

JunichiIto reviewed (approved):
1点コメントしましたが、すぐに直せる問題だと思うのでこれでOKです！

cellotak committed:
ls.rbの変数名を再度修正
